File: ios/Classes/Interstitial/AMPSInterstitialManager.swift
//
//  AMPSInterstitialManager.swift
//  amps_sdk
//

import Foundation
import Flutter
import AMPSAdSDK

class AMPSInterstitialManager: NSObject {

    static let shared = AMPSInterstitialManager()
    private override init() { super.init() }

    private var ads: [String: InterstitialAdEntry] = [:]

    private final class InterstitialAdEntry {
        let instanceId: String
        var interstitialAd: AMPSInterstitialAd?
        lazy var delegateHandler: InterstitialDelegateHandler = InterstitialDelegateHandler(entry: self)

        init(instanceId: String) {
            self.instanceId = instanceId
        }

        func cleanup() {
            interstitialAd?.delegate = nil
            interstitialAd = nil
        }
    }

    private final class InterstitialDelegateHandler: NSObject, AMPSInterstitialAdDelegate {
        private weak var entry: InterstitialAdEntry?

        init(entry: InterstitialAdEntry) {
            self.entry = entry
        }

        func ampsInterstitialAdLoadSuccess(_ interstitialAd: AMPSInterstitialAd) {
            guard let id = entry?.instanceId else { return }
            InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.onLoadSuccess, instanceId: id)
            InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.onRenderOk, instanceId: id)
        }

        func ampsInterstitialAdLoadFail(_ interstitialAd: AMPSInterstitialAd, error: (any Error)?) {
            guard let id = entry?.instanceId else { return }
            InstanceChannelHelper.send(
                AMPSAdCallBackChannelMethod.onLoadFailure,
                instanceId: id,
                data: ["code": (error as? NSError)?.code ?? 0, "message": (error as? NSError)?.localizedDescription ?? ""]
            )
        }

        func ampsInterstitialAdDidShow(_ interstitialAd: AMPSInterstitialAd) {
            guard let id = entry?.instanceId else { return }
            InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.onAdShow, instanceId: id)
        }

        func ampsInterstitialAdExposured(_ interstitialAd: AMPSInterstitialAd) {
            guard let id = entry?.instanceId else { return }
            InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.onAdExposure, instanceId: id)
        }

        func ampsInterstitialAdDidClick(_ interstitialAd: AMPSInterstitialAd) {
            guard let id = entry?.instanceId else { return }
            InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.onAdClicked, instanceId: id)
        }

        func ampsInterstitialAdShowFail(_ interstitialAd: AMPSInterstitialAd, error: (any Error)?) {
            guard let id = entry?.instanceId else { return }
            InstanceChannelHelper.send(
                AMPSAdCallBackChannelMethod.onAdShowError,
                instanceId: id,
                data: ["code": (error as? NSError)?.code ?? 0, "message": (error as? NSError)?.localizedDescription ?? ""]
            )
        }

        func ampsInterstitialAdDidClose(_ interstitialAd: AMPSInterstitialAd) {
            guard let id = entry?.instanceId else { return }
            InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.onAdClosed, instanceId: id)
            entry?.cleanup()
        }
    }

    func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case AMPSAdSdkMethodNames.interstitialLoad:
            handleInterstitialLoad(arguments: call.arguments, result: result)
        case AMPSAdSdkMethodNames.interstitialShowAd:
            handleInterstitialShowAd(arguments: call.arguments, result: result)
        case AMPSAdSdkMethodNames.interstitialGetEcpm:
            result(findEntry(from: call.arguments)?.interstitialAd?.eCPM() ?? 0)
        case AMPSAdSdkMethodNames.interstitialNotifyRtbWin:
            handleNotifyRTBWin(arguments: call.arguments, result: result)
        case AMPSAdSdkMethodNames.interstitialNotifyRtbLoss:
            handleNotifyRTBLoss(arguments: call.arguments, result: result)
        case AMPSAdSdkMethodNames.interstitialIsReadyAd:
            result(findEntry(from: call.arguments)?.interstitialAd != nil)
        default:
            result(false)
        }
    }

    private func findEntry(from arguments: Any?) -> InterstitialAdEntry? {
        guard let id = InstanceChannelHelper.instanceId(from: arguments) else { return nil }
        return ads[id]
    }

    private func handleInterstitialLoad(arguments: Any?, result: @escaping FlutterResult) {
        guard let map = arguments as? [String: Any],
              let instanceId = map[InstanceChannelKeys.instanceId] as? String else {
            result(false)
            return
        }

        ads[instanceId]?.cleanup()

        let entry = InterstitialAdEntry(instanceId: instanceId)
        let config = AdOptionModule.getAdConfig(para: map)
        entry.interstitialAd = AMPSInterstitialAd(adConfiguration: config)
        entry.interstitialAd?.delegate = entry.delegateHandler
        ads[instanceId] = entry
        entry.interstitialAd?.load()
        result(true)
    }

    private func handleInterstitialShowAd(arguments: Any?, result: @escaping FlutterResult) {
        guard let entry = findEntry(from: arguments),
              let interstitialAd = entry.interstitialAd,
              let vc = getKeyWindow()?.rootViewController else {
            result(false)
            return
        }
        interstitialAd.show(withRootViewController: vc)
        result(true)
    }

    private func handleNotifyRTBWin(arguments: Any?, result: @escaping FlutterResult) {
        guard let map = arguments as? [String: Any],
              let entry = findEntry(from: arguments) else {
            result(false)
            return
        }
        let winPrice = map[ArgumentKeys.adWinPrice] as? Int ?? 0
        let secPrice = map[ArgumentKeys.adSecPrice] as? Int ?? 0
        entry.interstitialAd?.sendWinNotification(withInfo: [BidKeys.winPrince: winPrice, BidKeys.lossSecondPrice: secPrice])
        result(true)
    }

    private func handleNotifyRTBLoss(arguments: Any?, result: @escaping FlutterResult) {
        guard let map = arguments as? [String: Any],
              let entry = findEntry(from: arguments) else {
            result(false)
            return
        }
        let lossWinPrice = map[ArgumentKeys.adWinPrice] as? Int ?? 0
        let lossSecPrice = map[ArgumentKeys.adSecPrice] as? Int ?? 0
        let lossReason = map[ArgumentKeys.adLossReason] as? String ?? ""
        entry.interstitialAd?.sendLossNotification(withInfo: [
            BidKeys.winPrince: lossWinPrice,
            BidKeys.lossSecondPrice: lossSecPrice,
            BidKeys.lossReason: lossReason,
        ])
        result(true)
    }
}

File: android/src/main/kotlin/com/example/amps_sdk/manager/AMPSInterstitialManager.kt
package com.example.amps_sdk.manager

import biz.beizi.adn.amps.ad.interstitial.AMPSInterstitialAd
import biz.beizi.adn.amps.ad.interstitial.AMPSInterstitialLoadEventListener
import biz.beizi.adn.amps.common.AMPSError
import biz.beizi.adn.amps.config.AMPSRequestParameters
import com.example.amps_sdk.data.*
import com.example.amps_sdk.utils.FlutterPluginUtil
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

/**
 * 多实例插屏广告管理器。
 */
class AMPSInterstitialManager private constructor() {

    private class AdEntry(val instanceId: String) {
        var interstitialAd: AMPSInterstitialAd? = null
    }

    private val ads = mutableMapOf<String, AdEntry>()

    companion object {
        @Volatile
        private var instance: AMPSInterstitialManager? = null

        fun getInstance(): AMPSInterstitialManager {
            return instance ?: synchronized(this) {
                instance ?: AMPSInterstitialManager().also { instance = it }
            }
        }
    }

    private fun findEntry(call: MethodCall): AdEntry? {
        val id = InstanceChannelHelper.instanceId(call.arguments) ?: return null
        return ads[id]
    }

    private fun getOrCreateEntry(instanceId: String): AdEntry {
        return ads.getOrPut(instanceId) { AdEntry(instanceId) }
    }

    private fun createAdListener(instanceId: String): AMPSInterstitialLoadEventListener =
        object : AMPSInterstitialLoadEventListener {
            override fun onAmpsAdLoaded() {
                InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.ON_LOAD_SUCCESS, instanceId)
                InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.ON_RENDER_OK, instanceId)
            }

            override fun onAmpsAdShow() {
                InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.ON_AD_SHOW, instanceId)
            }

            override fun onAmpsAdClicked() {
                InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.ON_AD_CLICKED, instanceId)
            }

            override fun onAmpsAdDismiss() {
                InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.ON_AD_CLOSED, instanceId)
                ads[instanceId]?.interstitialAd?.destroy()
                ads[instanceId]?.interstitialAd = null
            }

            override fun onAmpsAdFailed(error: AMPSError?) {
                InstanceChannelHelper.send(
                    AMPSAdCallBackChannelMethod.ON_LOAD_FAILURE,
                    instanceId,
                    mapOf("code" to error?.code, "message" to error?.message),
                )
            }

            override fun onAmpsSkippedAd() {
                InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.ON_VIDEO_SKIP_TO_END, instanceId)
            }

            override fun onAmpsVideoPlayStart() {
                InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.ON_VIDEO_PLAY_START, instanceId)
            }

            override fun onAmpsVideoPlayEnd() {
                InstanceChannelHelper.send(AMPSAdCallBackChannelMethod.ON_VIDEO_PLAY_END, instanceId)
            }
        }

    @Suppress("UNCHECKED_CAST")
    fun handleMethodCall(call: MethodCall, result: Result) {
        val args = call.arguments<Map<String, Any>?>()
        when (call.method) {
            AMPSAdSdkMethodNames.INTERSTITIAL_LOAD -> handleInterstitialLoad(call, result)
            AMPSAdSdkMethodNames.INTERSTITIAL_SHOW_AD -> handleInterstitialShowAd(call, result)
            AMPSAdSdkMethodNames.INTERSTITIAL_GET_ECPM -> {
                result.success(findEntry(call)?.interstitialAd?.ecpm ?: 0)
            }

            AMPSAdSdkMethodNames.INTERSTITIAL_NOTIFY_RTB_WIN -> {
                val entry = findEntry(call)
                val winPrice = args?.get(AD_WIN_PRICE) as? Number ?: 0
                val secPrice = args?.get(AD_SEC_PRICE) as? Number ?: 0
                entry?.interstitialAd?.notifyRTBWin(winPrice.toInt(), secPrice.toInt())
                result.success(null)
            }

            AMPSAdSdkMethodNames.INTERSTITIAL_NOTIFY_RTB_LOSS -> {
                val entry = findEntry(call)
                val lossWinPrice = args?.get(AD_WIN_PRICE) as? Number ?: 0
                val lossSecPrice = args?.get(AD_SEC_PRICE) as? Number ?: 0
                val lossReason =
                    args?.get(AD_LOSS_REASON) as? String ?: StringConstants.EMPTY_STRING
                entry?.interstitialAd?.notifyRTBLoss(
                    lossWinPrice.toInt(),
                    lossSecPrice.toInt(),
                    lossReason,
                )
                result.success(null)
            }

            AMPSAdSdkMethodNames.INTERSTITIAL_IS_READY_AD -> {
                result.success(findEntry(call)?.interstitialAd?.isReady ?: false)
            }

            else -> result.notImplemented()
        }
    }

    private fun handleInterstitialLoad(call: MethodCall, result: Result) {
        val activity = FlutterPluginUtil.getActivity()
        if (activity == null) {
            result.error("LOAD_FAILED", "Activity not available for loading interstitial ad.", null)
            return
        }

        val instanceId = InstanceChannelHelper.instanceId(call.arguments)
        if (instanceId.isNullOrEmpty()) {
            result.error("LOAD_FAILED", "instanceId is required.", null)
            return
        }

        try {
            val adOptionsMap = call.arguments<Map<String, Any>?>()
            val adOption: AMPSRequestParameters =
                AdOptionsModule.getAdOptionFromMap(adOptionsMap, activity)
            val entry = getOrCreateEntry(instanceId)
            entry.interstitialAd =
                AMPSInterstitialAd(activity, adOption, createAdListener(instanceId))
            entry.interstitialAd?.loadAd()
            result.success(true)
        } catch (e: Exception) {
            result.error("LOAD_EXCEPTION", "Error loading interstitial ad: ${e.message}", e.toString())
        }
    }

    private fun handleInterstitialShowAd(call: MethodCall, result: Result) {
        val entry = findEntry(call)
        val interstitialAd = entry?.interstitialAd
        val activity = FlutterPluginUtil.getActivity()
        if (interstitialAd == null) {
            result.error("SHOW_FAILED", "Interstitial ad not loaded.", null)
            return
        }
        if (activity == null) {
            result.error("SHOW_FAILED", "Activity not available.", null)
            return
        }
        interstitialAd.show(activity)
        result.success(true)
    }
}
